herbert_core/utilities/classes/@unique_references_container/unique_references_container.m

```diff
@@ -0,0 +1,571 @@
+            glc = self.global_container('value',self.global_name_);
+            [ix,~] = glc.find_in_container(obj);
+            inglc = ismember(ix, self.idx_);
+            if ~any(inglc)
+                obj_loc = [];
+            else
+                obj_loc = ix;
+            end
+        end
+    end
+    
+    methods (Access = protected) % get, add, replicate and replace
+        
+        % really only for use within class. these implement
+        % subsref/subsasgn action.
+        
+        % alternative access method
+        function val = get(self,index)
+            glc = self.global_container('value',self.global_name_);
+            val = glc{ self.idx(index) };
+        end
+        
+        % add a single object obj at the end of the container
+        function [self, nuix] = add_single_(self,obj)
+            if isempty(self.global_name_)
+                error('HERBERT:unique_references_container:incomplete_setup', ...
+                      'global name unset');
+            end
+            [glindex, ~] = self.global_container('value',self.global_name_).find_in_container(obj);
+            if isempty(glindex)
+                glcont = self.global_container('value',self.global_name_);
+                [glcont,glindex] = glcont.add(obj);
+                if glindex == 0
+                    % object was not added
+                    nuix = 0;
+                    return
+                end
+                self.global_container('reset',self.global_name_,glcont);
+            end
+            self.idx_ = [ self.idx(:)', glindex ];
+            nuix = numel(self.idx_);
+        end
+        
+        % add (possibly contents of multiple) objects at the end of the
+        % container
+        function [self, nuix] = add(self, obj)
+            if ~ischar(obj) && (numel(obj)>1 || iscell(obj) || ...
+                                isa(obj, 'unique_objects_container'))
+                nobj = 0;
+                if isa(obj, 'unique_objects_container')
+                    nobj = obj.n_runs;
+                elseif numel(obj)>1 || iscell(obj)
+                    nobj = numel(obj);
+                end
+                if nobj>0
+                    nuix = zeros(1,nobj);
+                    if iscell(obj)
+                        for i=1:nobj
+                            [self,nuix(i)] = self.add_single_(obj{i});
+                        end
+                    else
+                        for i=1:nobj
+                            [self,nuix(i)] = self.add_single_(obj(i));
+                        end
+                    end
+                    return;
+                end
+            end
+            if ~isa(obj,self.baseclass_)
+                warning('HERBERT:unique_references_container:invalid_argument', ...
+                        'not correct base class; object was not added');
+                nuix = 0;
+                return;
+            end
+            [self,nuix] = self.add_single_(obj);
```

Comment:
I find the logic here confusing and it would be a pain to debug. The following should work, but I haven't tested it:
```matlab
function [self, nuix] = add(self, obj)
    switch class(obj)
      case 'char'
        if ~isa(obj,self.baseclass_)
            warning('HERBERT:unique_references_container:invalid_argument', ...
                    'not correct base class (%s), found ''%s''; object was not added', ...
                    self.baseclass_, class(obj));
            nuix = 0;
            return;
        end
        nobj = size(a, 1);
        nuix = zeros(nobj, 1);
        for i=1:nobj
            [self,nuix(i)] = self.add_single_(obj(i,:));
        end

      case 'cell'
        if ~all(cellfun(@(x)isa(x, self.baseclass_), obj))
            warning('HERBERT:unique_references_container:invalid_argument', ...
                    'not all objects correct base class (%s); object was not added', ...
                    self.baseclass_);
            nuix = 0;
            return;
        end
        nobj = numel(obj);
        nuix = zeros(nobj, 1);
        for i=1:nobj
            [self,nuix(i)] = self.add_single_(obj{i});
        end

      case 'unique_objects_container'
        if ~strcmp(obj.baseclass,self.baseclass_)
            warning('HERBERT:unique_references_container:invalid_argument', ...
                    'not correct base class (%s), found ''%s''; object was not added', ...
                    self.baseclass_, obj.baseclass);
            nuix = 0;
            return;
        end
        nobj = obj.n_runs;
        nuix = zeros(nobj, 1);
        for i=1:nobj
            [self,nuix(i)] = self.add_single_(obj(i));
        end


      otherwise
        if ~isa(obj,self.baseclass_)
            warning('HERBERT:unique_references_container:invalid_argument', ...
                    'not correct base class (%s), found ''%s''; object was not added', ...
                    self.baseclass_, class(obj));
            nuix = 0;
            return;
        end
        nobj = numel(obj);
        nuix = zeros(nobj, 1);
        for i=1:nobj
            [self,nuix(i)] = self.add_single_(obj(i));
        end
    end
end
```

Reply:
I'm going to agree on the difficulty of understanding the code and disagree that case splitting helps - in fact case may obscure the logic. I have instead commented the code extensively, and fixed an issue in the process.